File: packages/java/parser-jvm-plugin-nonnull-kotlin/src/test/kotlin/com/vaadin/hilla/parser/plugins/nonnull/kotlin/basic/SimpleEndpoint.kt
package com.vaadin.hilla.parser.plugins.nonnull.kotlin.basic

import com.vaadin.hilla.parser.plugins.nonnull.kotlin.annotation.Endpoint
import com.vaadin.hilla.parser.plugins.nonnull.kotlin.basic.data.*
import jakarta.validation.constraints.Email
import jakarta.validation.constraints.NotBlank
import java.time.LocalDate


@Endpoint
open class SimpleEndpoint {

    fun sayHello(name: String?, age: Int): String =
        if (name?.isBlank() == true) {
            "Hello stranger"
        } else {
            "Hello $name"
        }

    fun isValidPerson(person: Person): Boolean? {
        return person.firstName.isNotBlank() && person.dateOfBirth != null
    }

    fun getPersonsByFirstname(firstname: String): List<Person> {
        return listOf()
    }

    fun saveAllPersons(persons: List<Person>): List<Person?>? {
        return listOf()
    }

    fun saveDepartmentManager(data: Map<String, Manager?>): Map<String?, Manager>? {
        return mapOf()
    }

    fun saveTeam(team: Team<Employee, Manager?>?): Team<Employee, Manager?>? {
        return team
    }

    fun saveTeams(teams: List<Team<Employee?, Manager>?>?): List<Team<Employee?, Manager>?> {
        return listOf()
    }

    fun saveCompaniesPerCities(data: Map<String, Map<String, List<Map<String, List<Team<Employee, Manager?>>?>>>>): Map<String, Map<String, List<Map<String, List<Team<Employee?, Manager>?>>>>?> {
        return mapOf()
    }

    fun saveNestedPerson(person: NestedPerson): NestedPerson {
        return person
    }

    class NestedPerson(
        var firstName: @NotBlank String = "",
        var lastName: String = "",
        var email: @Email String? = null,
        var phone: String = "",
        var address: MutableList<Address> = mutableListOf(),
        var team: Map<String, List<Employee?>> = mapOf(),
        var dateOfBirth: LocalDate? = null,
        var occupation: String? = null,
        var role: String? = null,
        var important: Boolean = false,
    )
}

File: packages/java/typescript-generator/src/test/resources/com/vaadin/hilla/parser/plugins/nonnull/kotlin/basic/snapshots/SimpleEndpoint.ts
import { EndpointRequestInit } from "@vaadin/hilla-frontend";
import type Employee from "./com/vaadin/hilla/parser/plugins/nonnull/kotlin/basic/data/Employee.js";
import type Manager from "./com/vaadin/hilla/parser/plugins/nonnull/kotlin/basic/data/Manager.js";
import type Person from "./com/vaadin/hilla/parser/plugins/nonnull/kotlin/basic/data/Person.js";
import type Team from "./com/vaadin/hilla/parser/plugins/nonnull/kotlin/basic/data/Team.js";
import type NestedPerson from "./com/vaadin/hilla/parser/plugins/nonnull/kotlin/basic/SimpleEndpoint/NestedPerson.js";
import client from "./connect-client.default.js";
async function getPersonsByFirstname(firstname: string, init?: EndpointRequestInit): Promise<Array<Person>> { return client.call("SimpleEndpoint", "getPersonsByFirstname", { firstname }, init); }
async function isValidPerson(person: Person, init?: EndpointRequestInit): Promise<boolean | undefined> { return client.call("SimpleEndpoint", "isValidPerson", { person }, init); }
async function saveAllPersons(persons: Array<Person>, init?: EndpointRequestInit): Promise<Array<Person | undefined> | undefined> { return client.call("SimpleEndpoint", "saveAllPersons", { persons }, init); }
async function saveCompaniesPerCities(data: Record<string, Record<string, Array<Record<string, Array<Team<Employee, Manager | undefined>> | undefined>>>>, init?: EndpointRequestInit): Promise<Record<string, Record<string, Array<Record<string, Array<Team<Employee | undefined, Manager> | undefined>>>> | undefined>> { return client.call("SimpleEndpoint", "saveCompaniesPerCities", { data }, init); }
async function saveDepartmentManager(data: Record<string, Manager | undefined>, init?: EndpointRequestInit): Promise<Record<string, Manager> | undefined> { return client.call("SimpleEndpoint", "saveDepartmentManager", { data }, init); }
async function saveNestedPerson(person: NestedPerson, init?: EndpointRequestInit): Promise<NestedPerson> { return client.call("SimpleEndpoint", "saveNestedPerson", { person }, init); }
async function saveTeam(team: Team<Employee, Manager | undefined> | undefined, init?: EndpointRequestInit): Promise<Team<Employee, Manager | undefined> | undefined> { return client.call("SimpleEndpoint", "saveTeam", { team }, init); }
async function saveTeams(teams: Array<Team<Employee | undefined, Manager> | undefined> | undefined, init?: EndpointRequestInit): Promise<Array<Team<Employee | undefined, Manager> | undefined>> { return client.call("SimpleEndpoint", "saveTeams", { teams }, init); }
async function sayHello(name: string | undefined, age: number, init?: EndpointRequestInit): Promise<string> { return client.call("SimpleEndpoint", "sayHello", { name, age }, init); }
export { getPersonsByFirstname, isValidPerson, saveAllPersons, saveCompaniesPerCities, saveDepartmentManager, saveNestedPerson, saveTeam, saveTeams, sayHello };
